assignment5/assignment5.py

```diff
@@ -0,0 +1,45 @@
+class AlienDictionary(object):
+    def alienOrder(self, words):
+        # Find ancestors of each node by DFS.
+        nodes, ancestors = sets.Set(), {}
+        for i in xrange(len(words)):
+            for c in words[i]:
+                nodes.add(c)
+        for node in nodes:
+            ancestors[node] = []
+        for i in xrange(1, len(words)):
+            if len(words[i-1]) > len(words[i]) and \
+                words[i-1][:len(words[i])] == words[i]:
+                    return ""
+            self.findEdges(words[i - 1], words[i], ancestors)
+
+        
+        result = []
+        visited = {}
+        for node in nodes:
+            if self.topSortDFS(node, node, ancestors, visited, result):
+                return ""
+        
+        return "".join(result)
+
+
+    # Find edges
+    def findEdges(self, word1, word2, ancestors):
+        min_len = min(len(word1), len(word2))
```

Comment:
Python has a helpful `zip` function that generates tuples with same-index elements of all input iterables up until the length of the shortest one. So:

```
for a, b in zip(word1, word2):
  if a != b:
    ancestors[b].append(a)
    break
```